src/app/loadout-builder/process/useProcess.ts

```diff
@@ -250,45 +255,113 @@ function groupItems(
   combatMods: PluggableInventoryItemDefinition[],
   activityMods: PluggableInventoryItemDefinition[]
 ) {
-  const groupingFn = (item: DimItem) => {
+  // We group in two passes - first by things like mod requirements, then by stats. In-between we drop any
+  // items which are strictly inferior in stats (all stats less than or equal to another item) from each group.
+
+  // Figure out all the energy types that have been requested across all mods
+  const requiredEnergyTypes = new Set<DestinyEnergyType>();
+  for (const mod of [...combatMods, ...generalMods, ...activityMods]) {
+    if (mod.plug.energyCost && mod.plug.energyCost.energyType !== DestinyEnergyType.Any) {
+      requiredEnergyTypes.add(mod.plug.energyCost.energyType);
+    }
+  }
+
+  // Figure out all the interesting mod slots required by mods are. This is just
+  // raid/nightmare, we don't care about combat and general mod slots because
+  // all items have them.
+  const requiredActivityModSlots = new Set<string>();
+  for (const mod of activityMods) {
+    const modTag = getModTypeTagByPlugCategoryHash(mod.plug.plugCategoryHash);
+    if (modTag) {
+      requiredActivityModSlots.add(modTag);
+    }
+  }
+
+  // Group by mod requirements (energy, slot). The groups are based on the mods
+  // requested - for example if we only request solar mods, the groups should be
+  // "solar", and "other", not "solar", "arc", "void", "stasis". If we only request
+  // a vault of glass mod, we don't make a group that includes deep stone crypt mods.
+  const modGroupingFn = (item: DimItem) => {
+    let groupId = '';
+
+    if (requiredActivityModSlots.size) {
+      const socketTags = getInterestingSocketMetadatas(item) ?? [];
+      for (const socketTag of socketTags) {
+        // Only add to the grouping key if the socket matches what we need
+        if (requiredActivityModSlots.has(socketTag.slotTag)) {
+          groupId += socketTag.slotTag;
+        }
+      }
+    }
+
+    if (requiredEnergyTypes.size) {
+      groupId +=
+        // Only add the grouping key if the socket matches what we need, otherwise it doesn't matter
+        item.energy &&
+        requiredEnergyTypes.has(item.energy.energyType) &&
+        // If we can swap to another energy type, there's no need to group by current energy type
+        !canSwapEnergyFromUpgradeSpendTier(defs, upgradeSpendTier, item, lockItemEnergyType)
+          ? item.energy.energyType
+          : DestinyEnergyType.Any;
+    }
+
+    return groupId;
+  };
+
+  const statsCache = new Map<DimItem, number[]>();
+  for (const item of items) {
     const statValues: number[] = [];
     const statsByHash = item.stats && keyByStatHash(item.stats);
     // Ensure ordering of stats
+    // TODO: statOrder includes disabled stats, should we omit them?
```

Comment:
I think so yes, if people do disable stats there are less unique vectors to compare by. Doing this will mean we send even less items to process. This can happen in a different PR if you want as this is already a huge improvement.